### ISSUE TITLE
--init number - not using proper number of threads to initialize

### ISSUE THREAD

**Issue by kio3i0j9024vkoenio** [tevador/RandomX#114]
System:

Dell R815 Server with quad 16-core 6378 Opteron's
64GB Memory (4x 4GB per Opteron)

Latest RandomX code:

git clone https://github.com/tevador/RandomX.git
cd RandomX
mkdir build && cd build
cmake -DARCH=native ..
make

I have two shellinabox's opened to the server. In one I monitor which cores are active with NMON. The other is where the RandomX test is done.

I test one Opteron 6378 at a time to check that there are any issues with that Opteron and it's local memory with this command:

seq 0 1 | xargs -P 0 -I node numactl -N node ./randomx-benchmark --mine --largePages --jit --nonces 10000 --init 8 --threads 8

seq 0 1 becomes seq 2 3 | seq 4 5 | seq 6 7 for the other Opteron's

This should initialize the DataSet on the selected Opteron (2 nodes per Opteron) and use 8 threads for initializing the DataSet in each node. 8 cores are in each NODE.

What I see doing the above command is only one thread (core) in each node being utilized for the initialization. The other seven cores are idle.

`RandomX benchmark v1.0.4                                                                                                                                                                     
 - full memory mode (2080 MiB)                                                                                                                                                               
 - JIT compiled mode                                                                                                                                                                         
 - hardware AES mode                                                                                                                                                                         
 - large pages mode                                                                                                                                                                          
Initializing (8 threads) ...                                                                                                                                                                 
RandomX benchmark v1.0.4                                                                                                                                                                     
 - full memory mode (2080 MiB)                                                                                                                                                               
 - JIT compiled mode                                                                                                                                                                         
 - hardware AES mode                                                                                                                                                                         
 - large pages mode                                                                                                                                                                          
Initializing (8 threads) ...                                                                                                                                                                 
Memory initialized in 42.7622 s                                                                                                                                                              
Initializing 8 virtual machine(s) ...                                                                                                                                                        
Running benchmark (10000 nonces) ...                                                                                                                                                         
Memory initialized in 43.5766 s                                                                                                                                                              
Initializing 8 virtual machine(s) ...                                                                                                                                                        
Running benchmark (10000 nonces) ...                                                                                                                                                         
Calculated result: 3412fd710459cb4e710e2f72760f632f4f00a0463ab02dbdd7dae176aeb1ccc2                                                                                                          
Performance: 1497.69 hashes per second                                                                                                                                                       
Calculated result: 3412fd710459cb4e710e2f72760f632f4f00a0463ab02dbdd7dae176aeb1ccc2                                                                                                          
Performance: 1499.56 hashes per second                                                                                                                                                       ` 

Note the time to initialize: Memory initialized in 42.7622 s

If I do --init 16 like this:

seq 0 1 | xargs -P 0 -I node numactl -N node ./randomx-benchmark --mine --largePages --jit --nonces 10000 --init 16 --threads 8

Then I get all 8 cores on each node being active in initialization and:
Memory initialized in 11.1164 s
or about four times faster.

A problem does arise with the above as the Performance drops:
Performance: 1330.81 hashes per second from Performance: 1499.56 hashes per second

If I try:

seq 0 1 | xargs -P 0 -I node numactl -N node ./randomx-benchmark --mine --largePages --jit --nonces 10000 --init 32 --threads 8

I again only get one core (thread) doing initialization and taking:
Memory initialized in 43.074 s
with the correct hash rate of:
Performance: 1494.93 hashes per second

There appears to be an issue with --init

**Comment by kio3i0j9024vkoenio** [comment]
Upon further testing I am getting different results when doing the same command:

seq 0 1 | xargs -P 0 -I node numactl -N node ./randomx-benchmark --mine --largePages --jit --nonces 100000 --init 8 --threads 8

Sometimes it only uses one thread (core) for initialization and takes 42 seconds and the next time it uses one core for a bit then uses all 8 cores for the rest of the initialization and only takes 11 seconds.

Any idea why?

**Comment by hyc** [comment]
At a guess, some other processes are tying up the particular NUMA memory regions and so preventing those CPUs from running.

**Comment by kio3i0j9024vkoenio** [comment]
It appears that one or more of the 6378 Opteron's has a defect as I have now tested without issues on another exactly same configured R815. That one has no issues what so ever.

On the one that is flaky I had already swapped known good (and tested) memory in it so memory is not an issue.

I will close this thread as the problems I am having is not the code but something in the system. If I find out what it is I will update in case someone else experiences it.
